### ISSUE TITLE
[HOLD for payment 2023-02-10] File name with  "/"  removes the extension and breaks the file when downloaded

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14266]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open Expensify
2. Open any report
3. Attach any file with "/" as the filename and valid file extension.
4. Download the file.

## Expected Result:
Valid file downloaded

## Actual Result:
Malformed file downloaded due to removed extension

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.53-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/212224807-7f129ec1-22aa-4a2c-b9f1-47ef3cd6f466.mov

https://user-images.githubusercontent.com/43996225/212225197-7fa16ef6-626a-4a4d-9f54-80ecb9d50c8c.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @oesayan
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1673509312939899

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01b3bd35954214a174</li>
        <li>Upwork Job ID: 1614013175418511360</li>
        <li>Last Price Increase: 2023-01-13</li>
    </ul>
</details>

**Comment by NicMendonca** [comment]
In the thread @mountiny mentioned this being internal, @dangrous please let me know otherwise!

**Comment by mountiny** [comment]
@dangrous i think it could have solution in both, but ideally we would make sure this is processed fine in the backend (havnet looked into where is the root cause of this)

**Comment by NicMendonca** [comment]
bump @dangrous 👋

**Comment by dangrous** [comment]
Hi sorry - was OOO yesterday! Will look into this today.

**Comment by dangrous** [comment]
Okay so figured out part of it -

When a user uploads an attachment, we sanitize the filename. This would turn, for example, `/.png` into just `.png` - we use this:
`preg_replace('/[^a-zA-Z0-9\-\._]/', '', $filename);`

The `/` actually gets converted into a `:` at some point in the process, so the regex also filters out that.

However, after that point I've confirmed that the filetype is correct, so I'm not sure why it's downloading as png.txt.

But either way, if we allow colons in the filename, the behavior is resolved. It downloads as `_.png` but you are able to open it fine. @mountiny what do you think about this? Is there an insecurity in allowing colons, or can you think of other behavior that might get messed up?

**Comment by mananjadhav** [comment]
@dangrous  I think one reason for not using `:` colons is Windows doesn't support `:` in the file names. So when saving the files the users would probably get an error. Are you saying that `:` gets converted into `_` underscore when the file is downloaded? Then that should be just fine.

**Comment by mountiny** [comment]
I cant really confidently answer if this opens door for some other issues.

I would say we could also not allow the `/` in file name, I dont think its very common to have filenames with this character.

But as Manan mentioned, I think converting it to underscore and running with that should be fine if we want just backend solution

**Comment by dangrous** [comment]
So at some point in the flow (I'm not sure where yet) the `/` gets converted into a `:` - this happens on the App side somewhere.

Then, the `:` is stripped out on the Web-E side.

After changing the regex in testing to allow `:`, somewhere between storing the file and downloading it, that `:` gets converted into a `_`.

So maybe if I can find where `/.png` becomes `:.png` on the App side, we could make it switch it into `_.png` and then we'd be all set. That could also be external, too.

**Comment by mananjadhav** [comment]
There's a lot of somewheres. So are we saying this could purely be external? with no change on the backend?

**Comment by mountiny** [comment]
I think we should do this in both, change the `/` to `_` in frontend and backend as well so we not only rely on the frontend and make sure we dont leave room for vulnerabilities.

**Comment by dangrous** [comment]
That's fair. So I can put up a PR that adjusts https://github.com/Expensify/Web-Expensify/blob/62f13038a4fe5e3f9991f1d2b23e7a00474a1827/lib/ReportUtils.php#L2221-L2224 to also replace any `/`s and `:`s with `_`s, and then we can open this externally to do the same on the App side?

For reference, I know it's getting switched to a `:` on the app side because when you attempt to upload a `/.png` the header says:
![image](https://user-images.githubusercontent.com/5487802/213298262-b5292c31-a034-4b54-9dd0-db268f02a6dd.png)
Before you hit the server at all.

**Comment by mananjadhav** [comment]
I think the `/` to `:` separator is done by the MacOS (*nix).

<img width="757" alt="image" src="https://user-images.githubusercontent.com/3069065/213297093-7f399fd2-1e85-44d2-acd1-e12bc10123d0.png">

and then when we get the file from the backend, we don't get any `/` or `:`.  The `originalMessage` content received from the backend.

```
<img src="https://www.expensify.com/chat-attachments/4313048699244575240/w_3249cb0371a0cac30c660b74ee79935ee9d0c9b0.png.1024.jpg" data-expensify-source="https://www.expensify.com/chat-attachments/4313048699244575240/w_3249cb0371a0cac30c660b74ee79935ee9d0c9b0.png" data-name=".png" data-expensify-width="225" data-expensify-height="224" />
```

Notice the `data-name` which only contains `.png` and hence we save the file it shows just png.

**Comment by dangrous** [comment]
Ah okay. So we'd need to add something there to switch it to a `_` but that should be fine. The data-name is only showing png because it goes through that regex above.

When I replace the regex on the backend to swap out `:` and `/`, it succesfully downloads from the front end as `[an underscore for each slash].png` so looks like it's working there, even without a frontend change.

**Comment by oesayan** [comment]
@mananjadhav is right - that's an old tale in MacOS world since before some old version MacOS used : instead of slashes and some layers of MacOS still use :, some layers use slash - Finder allows slashes to be used in filenames, while some other layers do not use slashes and replace them with : .
 That's why this behaviour is the same on iOS and MacOS, however on Android the OS replaces the slashes with _.

**Comment by mananjadhav** [comment]
I think @akshayasalvi's proposal is good here. One recommendation that, move the `.replace` call to `FileUtils`. 

> but if it needs to be done for the native too, I can do that.

Yes, let's add the check for native too.

@dangrous @mountiny What do you folks think?

**Comment by mountiny** [comment]
I think thats a good step and I think we should also explore updating this in backend to match the client behaviour.

Did we test this will fix the download issue?

**Comment by oesayan** [comment]
@mountiny So I tried playing around with it a little bit more and noticed the same behaviour if the file is named with emojis.

**Comment by akshayasalvi** [comment]
@mountiny I tested the code and it works fine with the download. 

But there's a **revised proposal** here. Because we can't change the name of a `File` object, we need to create another instance. So the lines change as:

```
file.uri = URL.createObjectURL(file);
const updatedFile = new File([file], file.name.replace(/[/:]/g, '_').replace(/_+/g, '_')); // new file with the new name
this.onPicked(updatedFile); // send the file with the new name

```

**Comment by mananjadhav** [comment]
Thanks for the revised proposal @akshayasalvi. I was just testing your revised changes and it seems to be working.


https://user-images.githubusercontent.com/3069065/213466152-6ec1f5f7-0f40-41a4-944c-27a1cfe2c58a.mov

cc - @mountiny

**Comment by mountiny** [comment]
Thanks for the discussion, I will let @dangrous handle it since he is assigned 😄

**Comment by dangrous** [comment]
I think this makes sense - but I know little about file pickers. This wouldn't end up taking double the space or anything, right? Either for us or for the user (are we creating a duplicate file on their device)?

Separately, two thoughts about the regex:
* What's the reasoning behind consolidating multiple `/`s and `:` into only one `_`? It's probably fine, but doesn't seem necessary.
* If we do want that, I think we can just do `replace(/[/:]+/g, '_')` vs. `replace(/[/:]/g, '_').replace(/_+/g, '_')` and get the same result

Let me know what you all think, and once we're confirmed on the direction (single underscore vs. multiple) I can make the PR for the backend, as well.

**Comment by oesayan** [comment]
@dangrous @mananjadhav 
Hey guys, I guess you missed my comment before. 
I am the guy who found the issue and I played with it around a little more and found the exact same behavior with emojis, arabic symbols and probably some other symbols too.

So you will probably need a regex not only changing the / and :, but everything except alphanumeric characters.

**Comment by akshayasalvi** [comment]
@dangrous You're right we can update the regex. 

1. The reason for multiple `/` or `:`, is that if we have a file: `:::.png` or `/:.png`, it would get renamed as `___.png` and `__.png`.

2. It doesn't create a duplicate file on the client side, just a temp file, because you can't update the name of uploaded file.

**Comment by dangrous** [comment]
re @oesayan - thanks for reposting, just confirmed! so maybe we want `replace(/[^a-zA-Z0-9\-\._]+/g, '_')` which would be everything not letters, numbers, `-`, `.`, and `_`. I think just for personal preference, let's do one `_` per symbol, so not group them. (I'll make the same adjustments for the backend PR.)

@mountiny you can't think of any reason why that would cause an issue, can you?

**Comment by akshayasalvi** [comment]
Okay I can work without grouping too. Then I just need to remove the second regex. I can also update the regex to the one you gave `replace(/[^a-zA-Z0-9\-\._]+/g, '_')` to just keep letters, numbers, and limited special characters.

Let me know @dangrous if any other changes are expected [here](https://github.com/Expensify/App/issues/14266#issuecomment-1397034620)

**Comment by mountiny** [comment]
@dangrous Nope, I think this should not break anything for sure.

Making this External as well since contributor will work on this issue too

**Comment by priyeshshah11** [comment]
@mountiny Is this issue looking for any new proposals or the above proposal has already been selected?

**Comment by mountiny** [comment]
Good question, thanks @priyeshshah11, I think we might be good here and apologies this was not made External earlier for fairer conditions. We havent been sure what is the best way forward.

@dangrous if you want more proposals though, feel free to add the `Help Wanted` label back

**Comment by dangrous** [comment]
Agh, sorry for the process confusion! I missed that this wasn't already External, since it was unlocked. I'll definitely make sure to check that everything's in the right state for future issues to make everything fairer!

I think for now though, the proposal put forward by @akshayasalvi makes sense, with the edits noted [here](https://github.com/Expensify/App/issues/14266#issuecomment-1401418686), I will assign you and we can move ahead with the PR.

I'll spin up the backend update as well; we don't need to hold on either one, but it'll be good to get them out relatively close to each other.

**Comment by dangrous** [comment]
The backend fix is deployed to staging! Soon to be on prod. @akshayasalvi how are we looking on the PR for the front end part of this?

**Comment by dangrous** [comment]
@akshayasalvi How are things going with the PR? We'd love to get that merged this week, and it seems like a pretty quick fix!

**Comment by akshayasalvi** [comment]
@dangrous It will be up in an hour.

**Comment by mananjadhav** [comment]
Currently testing this one.

**Comment by dangrous** [comment]
Still waiting on this to get pushed to prod but should be soon!

**Comment by Pujan92** [comment]
@dangrous @mananjadhav @akshayasalvi 

Associated PR seems to raise a regression where we won't able to reach out to Avatar Crop part and the infinite loader will remain on the screen after cleaning the filename.
Reason is `uri` will be undefined as the file object is overwritten here.

https://github.com/Expensify/App/blob/51f1a77e18f222806da32c5d6670d2de5267fa6f/src/components/AttachmentPicker/index.js#L39-L42

https://user-images.githubusercontent.com/14358475/216611655-e81806b3-9cb6-4670-b288-8aced226e179.mp4

To solve we need to make an order change

```diff
diff --git a/src/components/AttachmentPicker/index.js b/src/components/AttachmentPicker/index.js
index dddceb7cf7..dc6573e02c 100644
--- a/src/components/AttachmentPicker/index.js
+++ b/src/components/AttachmentPicker/index.js
@@ -36,10 +36,10 @@ class AttachmentPicker extends React.Component {
 
                         if (file) {
                             const cleanName = FileUtils.cleanFileName(file.name);
-                            file.uri = URL.createObjectURL(file);
                             if (file.name !== cleanName) {
                                 file = new File([file], cleanName);
                             }
+                            file.uri = URL.createObjectURL(file);
                             this.onPicked(file);
```

https://user-images.githubusercontent.com/14358475/216611547-b1dc061c-913d-4533-917d-17e4f3348759.mov


**Context for me** - https://github.com/Expensify/App/issues/14757

**Comment by dangrous** [comment]
Re regression - yes, it looks like this did cause a regression, based on @Pujan92's analysis. I also tried reverting the change locally and confirmed that behavior was correct prior to the change.

I'm not sure if a checklist change is needed; I think the thing that was missed here was honestly just proofreading - I should have noticed that the file would no longer have the uri if the name was changed. I guess technically we could have noticed the bug by testing all places where attachments were uploaded, rather than just as attachments - but I'm not sure that's the primary solution here.

**Comment by akshayasalvi** [comment]
@dangrous I just went through the comments. I didn't think upload Avatar will be any different as it is the same picker.  I can see another PR is merged, is there any action pending here for me?

**Comment by dangrous** [comment]
Hey @akshayasalvi so to clarify (this is confusing!) the step on the checklist above is actually to propose a regression test for the original bug - that is, that `/.png` etc. won't be able to be downloaded. I think it would be something along the lines of "Upload a file named ///.png and make sure you can download it." but formatted appropriately according to the link.

Does that make sense?

**Comment by dangrous** [comment]
Separately, @mananjadhav, I think this is likely just an edge case we didn't consider before, rather than a regression. The initial renaming looks like it was added 7/24/20 into the backend so it was quite some time ago (possibly before the App existed?). Therefore no PR to comment on, and I don't think we need the discussion, as long as we get the new regression test. Does that sound right to you?

**Comment by dangrous** [comment]
Bump @akshayasalvi and @mananjadhav on the above - thanks!

**Comment by akshayasalvi** [comment]
**Regression Test Proposal**

**Bug:** File name with special characters like "/" breaks the file download

**Proposed Test Steps:**
1. Go to URL https://staging.new.expensify.com/
2. Log in with any account & navigate to any conversation
3. Have an image/file with the name as : or / with the extension, for example, `///.png`.
4. In the selected conversation, upload the above file
5. Verify that the preview should show if it is an image and all occurrences of the special character '/' is replaced with `_`
6. Click on upload and ensure the file is successfully uploaded
7. Open the file in the attachment preview once uploaded
8. Ensure you're able to preview if it is an image
9. Click on download for the image/file and once the file is downloaded, it should open successfully.


@dangrous @mananjadhav  Is this okay for the test proposal?

**Comment by mananjadhav** [comment]
I was OOO and just getting back, waiting for @akshayasalvi to post the regression test.

**Comment by dangrous** [comment]
Hi @akshayasalvi this looks great! A couple edits - primarily I think we should just be very specific (you're right that it doesn't actually matter the kind of file, etc. but it'll be easier to test if it's particular).

1. Go to URL [staging.new.expensify.com](https://staging.new.expensify.com/).
2. Log in with any account & navigate to any conversation.
3. Have an image with the name `///.png`.
4. Upload the file to the conversation as an attachment.
5. Verify that the upload preview shows the image correctly, and the name of the file is shown as `___.png` (all occurrences of the special character `/` are replaced with `_`).
6. Click on upload and ensure the file is successfully uploaded.
7. Open the file in the attachment preview.
8. Ensure you're able to preview the image.
9. Download the image and once the file is downloaded (as `___.png`), ensure it opens successfully.

How does that sound?

**Comment by NicMendonca** [comment]
@mananjadhav @akshayasalvi @oesayan - can you three please apply to the job so I can issue payment? https://www.upwork.com/jobs/~01b3bd35954214a174

**Comment by akshayasalvi** [comment]
@NicMendonca applied, thank you.

@dangrous Yes this sounds good. It's my first time writing the regression test proposal. thank you for the help.

**Comment by mananjadhav** [comment]
@NicMendonca I've applied too.

**Comment by oesayan** [comment]
@NicMendonca and I have applied too 🤠

**Comment by NicMendonca** [comment]
@mananjadhav @akshayasalvi  - paid, thank you!

@oesayan - I'll issue payment once you accept the offer, thanks!

**Comment by NicMendonca** [comment]
everyone had been paid! thank you!!
